dot/parachain/backing/statement_table.go

```diff
@@ -422,33 +422,50 @@ type attestedCandidate struct {
 	validityAttestations []validatorIndexWithAttestation
 }
 
-func (attested *attestedCandidate) toBackedCandidate(tableCtx *tableContext) *parachaintypes.BackedCandidate {
-	group := tableCtx.groups[attested.groupID]
+func (attested *attestedCandidate) toBackedCandidate(tableCtx *tableContext) (*parachaintypes.BackedCandidate, error) {
+	if tableCtx == nil {
+		return nil, errors.New("table context is nil")
+	}
+
+	// Retrieve the group from tableContext
+	group, ok := tableCtx.groups[attested.groupID]
+	if !ok {
+		return nil, fmt.Errorf("validator group not found for the group-id: %d", attested.groupID)
+	}
+
+	// Create maps for validator index positions and validity votes
+	groupIndexMap := make(map[parachaintypes.ValidatorIndex]int)
+	for i, validator := range group {
+		groupIndexMap[validator] = i
+	}
+
 	validatorIndices := make([]bool, len(group))
-	var validityAttestations []parachaintypes.ValidityAttestation
-
-	// The order of the validity votes in the backed candidate must match
-	// the order of bits set in the bitfield, which is not necessarily
-	// the order of the `validity_votes` we got from the table.
-	for positionInGroup, validatorIndex := range group {
-		for _, validityVote := range attested.validityAttestations {
-			if validityVote.validatorIndex == validatorIndex {
-				validatorIndices[positionInGroup] = true
-				validityAttestations = append(validityAttestations, validityVote.validityAttestation)
-			}
+	validityVotes := make(map[int]parachaintypes.ValidityAttestation) // Map position in group to validity vote
+
+	// Separate ids and validity votes, and fill the map with the votes
+	for _, va := range attested.validityAttestations {
+		if pos, found := groupIndexMap[va.validatorIndex]; found {
+			validatorIndices[pos] = true
+			validityVotes[pos] = va.validityAttestation
+		} else {
+			return nil, errors.New("validity vote from unknown validator")
 		}
+	}
 
-		if !validatorIndices[positionInGroup] {
-			logger.Error("validity vote from unknown validator")
-			return nil
+	// Collect sorted validity votes
+	var sortedValidityVotes []parachaintypes.ValidityAttestation
+	for i := 0; i < len(group); i++ {
+		if vote, exists := validityVotes[i]; exists {
```

Comment:
If a vote doesn't exist at a position, this would change the order (index) in the resulting `sortedValidityVotes` slice.

Let's confirm if it is posible or okay for votes to not exists. It seems like it should not be okay.
If it is not okay,  throw an error if it doesnt exist.
If it is okay, add a nil vote to retain the order.

Reply:
As far as I have noticed.
The order needs to be the same and not the exact index.

also if we want to add a nil vote, the slice type must be `[]*ValidityAttestation`. current type is `ValidityVotes []ValidityAttestation`.
and we can't change the type because the scale encoding/decoding of `BackedCandidate` will change.

```
// BackedCandidate is a backed (or backable, depending on context) candidate.
type BackedCandidate struct {
	// The candidate referred to.
	Candidate CommittedCandidateReceipt `scale:"1"`
	// The validity votes themselves, expressed as signatures.
	ValidityVotes []ValidityAttestation `scale:"2"`
	// The indices of the validators within the group, expressed as a bitfield.
	ValidatorIndices BitVec `scale:"3"`
}
```